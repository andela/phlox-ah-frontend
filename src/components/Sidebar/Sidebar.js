import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import moment from 'moment';
import {
  Row, Col, Card, CardTitle
} from 'react-materialize';
import './Sidebar.scss';

let userPic = '';
let articlePic = '';
export const Sidebar = props => (
  <div>
    <h6 className=""><b>{props.sidebarTitle}</b></h6>
    { props.articles.map((article, index) => {// eslint-disable-line
      if (!article.User.Profile || !article.User.Profile.profileImage) {
        userPic = 'https://via.placeholder.com/300?text=AuthorsHaven';
      } else {
        userPic = article.User.Profile.profileImage;
      }
      if (!article.imgUrl) {
        articlePic = 'https://via.placeholder.com/300?text=AuthorsHaven';
      } else {
        articlePic = article.imgUrl;
      }
      return (<Row key={ index + 1 }>
        <Col s={12} m={12}>
          <Card className="z-depth-1" horizontal header={<CardTitle image={articlePic}></CardTitle>}>
              <h6><b><Link key={0} to={`/articles/${article.slug}`}>{article.title.substring(0, 25)}</Link></b></h6>
              <p>{article.description.substring(0, 50)}.<br/>
                </p>
                <Row className="author-details valign-wrapper">
                  <Col s={12} l={3} className="center-align">
                    <img className="img-responsive circle" src={userPic}/>
                  </Col>
                  <Col s={12} l={9}>
                    <p className="text-darken-4"><span><b>{article.User.username}</b></span> <span>{moment(article.createdAt).format('MMMM D YYYY')}</span></p>
                  </Col>
                </Row>
            </Card>
        </Col>
      </Row>);
    }) }
  </div>
);

Sidebar.propTypes = {
  articles: PropTypes.array,
  sidebarTitle: PropTypes.string,
};

export default Sidebar;
