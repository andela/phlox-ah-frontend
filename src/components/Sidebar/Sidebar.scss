.sidebar {
  a {
    cursor: pointer;
  }
  width: 29%;
  padding: 15px;
  h6 {
    color: #6F6A6A;
  }
  .card {
    height: 105px;
    box-shadow: none;
    .card-image {
      width: 35%;
      max-width: 35%;
      img {
        width: 100%;
      }
    }
    img {
      height: 100%;
      width: 100%;
    }
    .card-content {
      padding: 0px 14px 0px 14px;
      p {
        font-size: 0.7em;
        color: #ADA9A9;
        text-align: justify;
        line-height: 15px;
        word-spacing: -1px;
        span {
          font-color: black;
        }
      }
      h6 {
        font-size: 0.8em;
        margin-top: 2px;
        color: #5E5F63;
        a {
          color: #5E5F63;
          text-decoration: none;
          &:hover {
            color: #5E5F63;
          }
        }
      }
      .author-details {
        img {
          margin: auto;
          width: 20px;
          height: 20px;
        }
        .sb-avatar {
          width: 20px;
          height: 20px;
        }
        p {
          line-height: 30px;
          font-size: 0.8em;
          word-spacing: 0px;
        }
        span {
          display: block;
          font-size: 0.8em;
          color: #6F6A6A;
          &:nth-of-type(2) {
            margin-top: -15px;
          }
        }
      }
    }
  }
}
.card-stacked {
  max-width: 65%;
}
