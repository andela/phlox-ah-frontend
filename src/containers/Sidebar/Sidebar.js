import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Avatar from 'react-avatar';
import { Link } from 'react-router-dom';
import moment from 'moment';
import {
  Row, Col, Card, CardTitle, Badge
} from 'react-materialize';
import './Sidebar.scss';
import { getArticles } from '../../requests/ArticleRequests';
import { getAllTags } from '../../requests/TagRequests';

let authorAvatar = '';
let articlePic = '';

/**
 * @class Sidebar
 * @extends {Component}
 */
class Sidebar extends Component {
  /**
   *Creates an instance of Home.
   * @param {object} props
   * @memberof Sidebar
   */
  constructor(props) {
    super();
    this.state = {
      sidebarArticles: [],
      failure: true,
      success: false,
      tags: []
    };
  }

  /**
   *
   * @returns {func} tag
   * @memberof Sidebar
   */
  componentDidMount() {
    this.props.getArticles();
    this.props.getAllTags();
  }

  /**
   * @description - This method runs whenever redux state changes
   * @returns {object} state
   * @param {object} props
   * @param {object} state
   * @memberof Sidebar
   */
  static getDerivedStateFromProps(props, state) {
    return {
      tags: props.tags,
      failure: props.failure,
      sidebarArticles: props.sidebarArticles,
      success: props.success
    };
  }

  /**
   *
   * @returns {jsx} - jsx
   * @memberof Sidebar
   */
  showSidebarArticles() {
    if (this.state.failure || !this.state.sidebarArticles) {
      return (<div className="preloaderDiv"></div>);
    }
    return this.state.sidebarArticles.slice(0, 4).map((article, index) => {// eslint-disable-line
      if (!article.User.Profile || !article.User.Profile.profileImage) {
        authorAvatar = <Avatar name={article.User.username} size="20" round={true} />;
      } else {
        authorAvatar = <img className="img-responsive circle" src={article.User.Profile.profileImage}/>;
      }
      if (!article.imgUrl) {
        articlePic = 'https://via.placeholder.com/300?text=AuthorsHaven';
      } else {
        articlePic = article.imgUrl;
      }
      return (<Row key={ index + 1 }>
        <Col s={12} m={12}>
          <Card className="z-depth-1" horizontal header={<CardTitle image={articlePic}></CardTitle>}>
              <h6 className="capitalize"><b> <Link key={0} className="capitalize" to={`/articles/${article.slug}`}>{article.title.substring(0, 25)}</Link> </b></h6>
              <p>{article.description.substring(0, 45)}.<br/>
                </p>
                <Row className="author-details valign-wrapper">
                  <Col s={12} l={3} className="center-align">
                    { authorAvatar }
                  </Col>
                  <Col s={12} l={9}>
                    <p className="text-darken-4"><span><b>{article.User.username}</b></span> <span>{moment(article.createdAt).format('MMMM D YYYY')}</span></p>
                  </Col>
                </Row>
            </Card>
        </Col>
      </Row>);
    });
  }

  /**
   *
   *
   * @param {*} x
   * @returns {jsx} - jsx
   * @memberof Sidebar
   */
  showTags() {
    if (this.state.tags.length > 0) {
      return this.state.tags.map((tag, index) => {// eslint-disable-line
        return (<div key={index}>
        <Link to={`/tags/${tag.name}`}><Badge className="cyan">{ tag.name }</Badge></Link></div>);
      });
    }
    return (<div>No tags yet!</div>);
  }

  /**
   * @description - This method renders the jsx for this component
   * @returns {jsx} - jsx
   * @memberof Sidebar
   */
  render() {
    return (
      <div>
        <h6 className=""><b>Latest Articles</b></h6>
        { this.showSidebarArticles() }
        {/* <span className="title center">Tags</span> */}
        <h6 className=""><b>Tags</b></h6>
        <div className="sidebar-tag-container">
          { this.showTags() }
        </div>
      </div>);
  }
}

const mapStateToProps = state => ({
  tag: state.tags.tag,
  tags: state.tags.tags,
  failure: state.tags.failure,
  sidebarArticles: state.article.articles,
  success: state.tags.success,
});


Sidebar.propTypes = {
  articles: PropTypes.array,
  failure: PropTypes.bool,
  getAllTags: PropTypes.func,
  getArticles: PropTypes.func,
  getTag: PropTypes.func,
  sidebarArticles: PropTypes.array,
  sidebarTitle: PropTypes.string,
  success: PropTypes.bool,
};

export default connect(mapStateToProps, {
  getAllTags, getArticles
})(Sidebar);
