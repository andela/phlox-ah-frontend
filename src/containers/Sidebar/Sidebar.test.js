import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import Sidebar from './Sidebar';

const mockStore = configureMockStore();
const store = mockStore({
  info: {
    message: ['password is required'],
    success: false
  },
  tags: {
    error: true,
    tags: [],
    article: {
      articles: {
        title: ''
      }
    }
  },
  article: {
    articles: []
  },
  getArticles: jest.fn()

});

let component;

const props = {
  match: {
    url: '/tags/sports',
  },

};

describe('<Sidebar/>', () => {
  beforeEach(() => {
    component = shallow(
      <Provider store={store}>
        <Sidebar {...props} />
      </Provider>
    );
  });
  it('should render without throwing an error', () => {
    expect(component).toMatchSnapshot();
  });
});