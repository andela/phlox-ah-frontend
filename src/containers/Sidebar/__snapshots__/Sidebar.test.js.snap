// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar/> should render without throwing an error 1`] = `
<Connect(Sidebar)
  match={
    Object {
      "url": "/tags/sports",
    }
  }
/>
`;
